parlai/core/torch_generator_agent.py

```diff
@@ -1750,27 +1728,35 @@ def select_paths(self, logprobs, prior_scores, current_length) -> _PathSelection
         # get the actual word id from residual of the same division
         tok_ids = best_idxs % voc_size
 
-        tok_scores, tok_ranks = None, None
+        token_details: Optional[List[_PathSelectionTokenDetails]] = None
         if self.verbose:
-            tok_scores = (
-                torch.index_select(logprobs, 0, hyp_ids)
+            probs = torch.softmax(logprobs, dim=-1)
+            tok_probs = (
+                torch.index_select(probs, 0, hyp_ids)
                 .gather(1, tok_ids.unsqueeze(1))
                 .view(-1)
             )
-
             tok_ranks = (
-                logprobs.argsort(1, descending=True)
+                probs.argsort(1, descending=True)
                 .argsort(1)
                 .view(-1)
                 .gather(0, best_idxs)
             )
 
+            token_details = []
+
+            for tok_prob, tok_rank in zip(
+                tok_probs.cpu().numpy(), tok_ranks.cpu().numpy()
```

Comment:
.numpy May implicitly cpu?

Reply:
Seems to be a best practice to ensure behavior is the same when tensor may or may not be on GPU: https://discuss.pytorch.org/t/should-it-really-be-necessary-to-do-var-detach-cpu-numpy/35489/6